File: Java/dp/ClimbingStairs.java
/**
Given that a person can climb stairs in
steps of 1 and 2, find in how many ways the 
person can reach the top stair
2 2 1
1 2 2
are different sequences i.e., relative ordering 
matters
**/

import java.io.*;
import java.util.*;

public class ClimbingStairs {

    private static int climbStairs(int n) {
        if(n < 3)
            return n == 2 ? n : 1;
        int[] arr = new int[n+1];
        arr[0] = 1;
        arr[1] = 1;
        arr[2] = 2;
        // compute the dp table
        for(int i = 3; i <= n; i++)
            arr[i] = arr[i-1] + arr[i-2];
        return arr[n];
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        System.out.println(climbStairs(n));
    }
}

/**

Input :
3
Output :
3
Explanation :
1 1 1
1 2
2 1

Time Complexity : O(n)
Space Complexity : O(n)

**/

File: JavaScript/dp/Climbing_stairs.js
/*
Problem statement : 
A Person is at the bottom of n stairs .
To reach the top he can climb either 1 or 2 stairs at a time. 
Find the number of ways he can reach the top
*/

function climbing_stairs(n) {
    let dp = new Array(n);
    dp[0] = 1
    dp[1] = 1
    for (let i = 2; i <= n; i++) {
        dp[i] = dp[i - 1] + dp[i - 2];
    }
    return dp[n];
}
//For reading input 
const readline = require('readline');

const rl = readline.createInterface({ input: process.stdin, output: process.stdout });

const getLine = (function () {
    const getLineGen = (async function* () {
        for await (const line of rl) {
            yield line;
        }
    })();
    return async () => ((await getLineGen.next()).value);
})();

const main = async () => {
    console.log("Enter the number stairs");
    let n = Number(await getLine());
    console.log("The number of ways to climb the stairs is : " + climbing_stairs(n));
    process.exit(0);
};

main();

/*
Sample I/O:

Enter the number stairs
5
The number of ways to climb the stairs is : 8

Time complexity : O(n)
Space complexity : O(n)
*/
